### DISCUSSION TITLE
Is possible to change the the editor for a textfield by code?

### DISCUSSION THREAD

**Discussion by AmadeuAntunes** [OrchardCMS/OrchardCore#8681] [General]
I would like to change the editor for a textfield before render it.

**Comment by jptissot** [comment]
You can in a migration file. What do you mean when you say "Before render it"?

**Reply by AmadeuAntunes** [reply]
before use `@await this.DisplayAsync(myshape)`

I would like to switch the editor of a ContentParts by code 

![image](https://user-images.githubusercontent.com/47931101/109310867-fc3a9280-783c-11eb-924a-35961e6ca3a3.png)

**Reply by willnationsdev** [reply]
@AmadeuAntunes I might recommend that you create a shape template with a shape type name targeting the Edit display type of that type/part/field so that you can make it render a hidden input with your hardcoded value instead of a select dropdown. See the [Templates documentation](https://docs.orchardcore.net/en/dev/docs/reference/modules/Templates/) for examples of template names and how they match up with shape type to override specific types of content.

**Reply by AmadeuAntunes** [reply]
I tried this way: 
![image](https://user-images.githubusercontent.com/47931101/109497565-21701080-7a8a-11eb-982d-25e06fad6900.png)

and in template I used 
![image](https://user-images.githubusercontent.com/47931101/109497621-3d73b200-7a8a-11eb-92c8-8b57b66c6e19.png) 
this way I could change the html render, however it affect all fields use this, 

1 - is there other way to communicate with field template render without use ViewData?

2 - is there any way specify the field I would like to change readonly (if I use more then one field with this template using the viewData will change it for all)?

**Reply by AmadeuAntunes** [reply]
![image](https://user-images.githubusercontent.com/47931101/109514806-51c1aa00-7a9e-11eb-9e6b-1eb4b5237d98.png)
![image](https://user-images.githubusercontent.com/47931101/109514887-64d47a00-7a9e-11eb-9bf6-b881f4e4fdc1.png)
 I just solved this way, is there a better solution ?

**Reply by willnationsdev** [reply]
If you're concerned about the template conflicting with other fields of the same type, you can just give the field a specific name and then use that, I think. At least, that's what the [Templates docs](https://docs.orchardcore.net/en/dev/docs/reference/modules/Templates/) seem to indicate, in their examples where the `[FieldName]` is part of the template, like this:

```
[PartType]__[FieldName]
[ContentType]__[PartName]__[FieldName]
[PartType]__[FieldName]__[ShapeType]
[ContentType]__[PartName]__[FieldName]__[ShapeType]
```

But, now that I look at your code, it seems like you're trying to define a piece of Settings data that gets associated with each instance of the given field type. I guess I'm confused why you don't just try to create a Settings value similar to the TextFieldSettings. If you wanted a module to add a `ReadOnly` checkbox to the settings fro the text field, you could probably do that. So you'd end up with something along the lines of...

```cshtml
@{
    var readOnlySettings = Model.PartFieldDefinition.GetSettings<ReadOnlySettings>();
    var readOnlyAttribute = readOnlySettings.IsReadOnly ? "readonly " : "";
}
<input asp-for="Text" class="form-control content-preview-text" @readOnlyAttribute dir="...etc.">
```
In an example such as this, every instance of the field gets its own copy of the Settings object and the data gets merged as part of the PartFieldDefinition. The only thing you have to do is define the settings and implement DisplayDriver that handles the Display/Edit/Update logic for the Settings object and the .cshtml template that defines the rendering logic.

For examples on *how* to do this, I would look at the OrchardCore source code and how it defines and updates TextFieldSettings.

**Reply by Skrypt** [reply]
If you want a readonly TextField just create an editor for it.

If you look at the OrchardCore.ContentFields module you will see how we define these editors.
You basically just need to create a TextField-ReadOnly.Option.cshtml file so that this "shape template" be added to the list of editor options. And then you create a TextField.ReadOnly.Edit.cshtml file with your custom textfield html. No need for a special field setting to do that.

Put that inside a module that extends the OrchardCore.ContentFields module and it should work.
